### DISCUSSION TITLE
PatchGenerator does not pad

### DISCUSSION THREAD

**Discussion by mhoibo** [FAST-Imaging/FAST#178] [General]
I have noticed that patches generated with fast.PatchGenerator on edge cases are not padded. I have padded them manually, but I was wondering if the generator should do it automatically? paddingValue is set to -1 by default, which I though should mean that the patches get padded?

**Comment by smistad** [comment]
The PatchGenerator should always pad, even when paddingValue is -1, this only means that it should auto select of the padding is white (255) or black (0). See here: https://github.com/smistad/FAST/blob/master/source/FAST/Algorithms/ImagePatch/PatchGenerator.cpp#L150

If you are observing something else, it must be a bug..

**Reply by mhoibo** [reply]
thats what i though, but it does not happen when i use the patchgenerator. I'll try and figure out what I have done wrong:)

**Reply by andreped** [reply]
As far as I could tell, something wrong with padding in the PatchGenerator is the likely cause, as the result looked distorted and misplaced along the bottom-right edges after stitching.

In pyFAST, we fixed the issue by adding a simple Python PO that padded for us, which fixed the distortion. Note that @mhoibo was using `pyFAST==4.6.0`. Might be that this has been resolved in the latest release.

Example PO that was used (class name == :fire:):
```
import fast
import numpy as np

class McPaddy(fast.PythonProcessObject):
    def __init__(self, width=1024, height=1024):
        super().__init__()
        self.createInputPort(0)
        self.createOutputPort(0)
        self.height = height
        self.width = width

    def execute(self):
        image = self.getInputData()
        np_image = np.asarray(image)
        tmp = np.zeros((self.height, self.width, 3), dtype="uint8")
        shapes = np_image.shape
        tmp[:shapes[0], :shapes[1]] = np_image

        # Create new fast image and add as output
        new_output_image = fast.Image.createFromArray(tmp)
        new_output_image.setSpacing(image.getSpacing())
        self.addOutputData(0, new_output_image)
```

**Reply by smistad** [reply]
I added a test in FAST which checks that all patches coming from the PatchGenerator has the same size. The test passes fine, so the patches are padded at the edges.

**Reply by andreped** [reply]
I assume it is [this](https://github.com/smistad/FAST/blob/master/source/FAST/Algorithms/ImagePatch/Tests.cpp#L17) test you mean, @smistad?

@mhoibo is using an artifact of `pyFAST==4.6.0`, hence, this might have been fixed a while back.

But in @mhoibo case, she is having this problem when she has converted a numpy array to a FAST image and then streamed patches from it. Could there be a reason why this would not work then? She is also using an `overlapPercent=0.25`, but it seems to be an issue regardless if she performs overlap or not.

I can make a simple gist to see if I can reproduce the issue and share it here.

**Reply by smistad** [reply]
Okay, I see. When you get patches from an image pyramid it seems to pad, but not when getting patches from a regular image.

**Reply by andreped** [reply]
```
import fast
import numpy as np

x = np.zeros((1256, 1010), dtype="float32")
image = fast.Image.createFromArray(x)
patchgen = fast.PatchGenerator.create(256, 256, overlapPercent=0).connect(0, image)
streamer = fast.DataStream(patchgen)
for patch in streamer:
    print(np.asarray(patch).shape)
```

For most patches, we get `(256, 256, 1)`, but we occationally get patches of smaller size. Hence, padding does not seem to work appropriately in this case.

Just tried to do the same with latest stable release of pyFAST in a new virtual environement. So I believe the issue is coming from using the FAST image converted from a numpy array as input to the `PatchGenerator`.

**Reply by andreped** [reply]
> Okay, I see. When you get patches from an image pyramid it seems to pad, but not when getting patches from a regular image.

Yes, exactly. It is likely the issue. You can use the example above to test it for yourself.

**Reply by smistad** [reply]
This should be fixed in latest commit
